From: Valentine Fatiev <valentinef@nvidia.com>
Subject: [PATCH] BACKPORT: drivers/vdpa/mlx5/core/mr.c

---
 drivers/vdpa/mlx5/core/mr.c | 2 ++
 1 file changed, 2 insertions(+)

--- a/drivers/vdpa/mlx5/core/mr.c
+++ b/drivers/vdpa/mlx5/core/mr.c
@@ -1,6 +1,7 @@
 // SPDX-License-Identifier: GPL-2.0 OR Linux-OpenIB
 /* Copyright (c) 2020 Mellanox Technologies Ltd. */
 
+#ifdef HAVE_VDPA_SUPPORT
 #include <linux/vhost_types.h>
 #include <linux/vdpa.h>
 #include <linux/gcd.h>
@@ -619,3 +620,4 @@ int mlx5_vdpa_handle_set_map(struct mlx5
 
 	return err;
 }
+#endif
