From: Valentine Fatiev <valentinef@nvidia.com>
Subject: [PATCH] BACKPORT: drivers/vfio/pci/mlx5/main.c

---
 drivers/vfio/pci/mlx5/main.c | 4 ++++
 1 file changed, 4 insertions(+)

--- a/drivers/vfio/pci/mlx5/main.c
+++ b/drivers/vfio/pci/mlx5/main.c
@@ -3,6 +3,7 @@
  * Copyright (c) 2021-2022, NVIDIA CORPORATION & AFFILIATES. All rights reserved
  */
 
+#ifdef HAVE_VFIO_SUPPORT
 #include <linux/device.h>
 #include <linux/eventfd.h>
 #include <linux/file.h>
@@ -1320,7 +1321,9 @@ static const struct vfio_device_ops mlx5
 	.bind_iommufd = vfio_iommufd_physical_bind,
 	.unbind_iommufd = vfio_iommufd_physical_unbind,
 	.attach_ioas = vfio_iommufd_physical_attach_ioas,
+#ifdef HAVE_DETACH_IOAS_NDO
 	.detach_ioas = vfio_iommufd_physical_detach_ioas,
+#endif
 };
 
 static int mlx5vf_pci_probe(struct pci_dev *pdev,
@@ -1375,6 +1378,7 @@ static struct pci_driver mlx5vf_pci_driv
 };
 
 module_pci_driver(mlx5vf_pci_driver);
+#endif /* HAVE_VFIO_SUPPORT */
 
 MODULE_IMPORT_NS(IOMMUFD);
 MODULE_LICENSE("GPL");
